x-pack/plugins/fleet/server/services/actions/actions.ts

```diff
@@ -0,0 +1,252 @@
+import { auditLoggingService } from '../audit_logging';
+
+import { appContextService } from '../app_context';
+
+import type { FleetAction, FleetActionResponse } from './types';
+
+export const createAction = async (
+  esClient: ElasticsearchClient,
+  action: Partial<FleetAction>
+): Promise<Partial<FleetAction>> => {
+  try {
+    const body = {
+      ...action,
+      action_id: action.action_id ?? uuidV4(),
+      '@timestamp': action['@timestamp'] ?? new Date().toISOString(),
+    };
+    await esClient.create(
+      {
+        index: AGENT_ACTIONS_INDEX,
+        id: body.action_id,
+        body,
+        refresh: 'wait_for',
+      },
+      {
+        meta: true,
+      }
+    );
+
+    auditLoggingService.writeCustomAuditLog({
+      message: `User created Fleet action [id=${action.action_id}]`,
+    });
+
+    return body;
+  } catch (createActionError) {
+    throw new FleetActionsError(
+      `Error creating action: ${createActionError.message}`,
+      createActionError
+    );
+  }
+};
+
+type BulkCreate = Array<{ create: { _index: string; _id: string } } | object>;
+
+export const bulkCreateActions = async (
+  esClient: ElasticsearchClient,
+  _actions: Array<Partial<FleetAction>>
+): Promise<
+  Array<{
+    id: string;
+    action: Partial<FleetAction>;
+  }>
+> => {
+  const actions: Array<Partial<FleetAction>> = [];
+  const bulkCreateActionsBody = _actions.reduce<BulkCreate>((acc, action) => {
+    acc.push({ create: { _index: AGENT_ACTIONS_INDEX, _id: action.action_id ?? uuidV4() } });
+    const actionDoc = {
+      ...action,
+      action_id: action.action_id ?? uuidV4(),
+      '@timestamp': action['@timestamp'] ?? new Date().toISOString(),
+    };
+    acc.push(actionDoc);
+    actions.push(actionDoc);
+
+    return acc;
+  }, []);
+
+  try {
+    const bulkCreateActionsResponse = (await esClient.bulk(
+      {
+        body: bulkCreateActionsBody,
+        refresh: 'wait_for',
+      },
+      {
+        meta: true,
+      }
+    )) as unknown as estypes.BulkResponse;
+
+    const [errorItems, successItems] = partition(
+      bulkCreateActionsResponse.items,
+      (a) => a.create?.error
+    );
+
+    successItems.forEach((successItem) => {
+      auditLoggingService.writeCustomAuditLog({
+        message: `User created Fleet action [id=${successItem.create?._id}]`,
+      });
+    });
+
+    // log errored items
+    if (errorItems.length) {
+      appContextService.getLogger().error(new FleetActionsError(JSON.stringify(errorItems)));
+    }
+
+    // return success items
+    return successItems.map((item, i) => ({
+      id: item.create!._id as string,
+      action: actions[i],
+    }));
+  } catch (createBulkActionsError) {
+    throw new FleetActionsError(
+      `Error creating bulk actions: ${createBulkActionsError.message}`,
+      createBulkActionsError
+    );
+  }
+};
+
+export const getActionsByIds = async (
+  esClient: ElasticsearchClient,
+  actionIds: string[]
+): Promise<FleetAction[]> => {
```

Comment:
I don't think this return type if correct, because some IDs may not match any documents.


Maybe the return type should be:

```typescript
Promise<Array<{
  id: string; // action id requested (from 'actionIds' argument)
  action: FleetAction | undefined;
}>>
```